woocommerce/woocommerce-admin
# Add abtest package and tracks events to validate

This adds an A/B Test package to be used for WCPay experiments. 

In order to validate the randomness of the groups, I've added the following tracks events:

```
wcadmin_abtest_assign_variant

props:
abtest_name
abtest_group
context

wcadmin_abtest_serve_variant

props:
abtest_name
abtest_group
context
```

We've also set up a test in the payment task page when WCPay is a visible method, but with no variation (control = experiment) just so we can validate our randomization approach. I've added the test here as this is where we plan to run our first experiment (see https://wp.me/paJDYF-15q)

~~A note on storing variations in `wp_options`: Currently, in order to persist the assigned group across browsers/different admins, I am storing this as an option. I have not added any code to clean this up as we will only be running one test at a time and this is something we can manually take care of when finishing up one test/starting the next. This is not an ideal approach, so if there is a better one, let me know!~~

### Detailed test instructions:

Before testing, it is a good idea to replace the ABTest experiment prop in `client/task-list/tasks/payments/index.js ` with something easily identifiable. So something like 

```
<ABTest
	name ="validate_randomization_mechanism"
	control={ card }
	experiment={ <div>Hello Dolly</div> } <<< This line here.
	onComplete={ this.handleABTestComplete }
	key={ key }
/>
```

-  Go to the [payment task page](https://cloudup.com/cewDGNtBBCR) (making sure your test store is eligible for WC Pay)
- This should assign you a group, then save to both localStorage (`validate_randomization_mechanism` key) and wp_options table (`woocommerce_admin_abtests`) Confirm these groups match and the payment task reflects this group correctly.
- Delete the group in localStorage, then refresh the page. Confirm the same group is stored in localStorage once again and the payment task reflects this.
- Delete the option in wp_options, then refresh the page. Confirm the payment task reflects the correct group. Note, we do not re-save the option in this case.
- Delete both the localStorage key and the option, then refresh the page. Do this a few times to confirm you are assigned a different group than the first time.
- Finally, check [Tracks Live View](https://mc.a8c.com/tracks/live/?eventname=%25wcadmin_abtest%) for associated events and ensure they match what is expected. There should be only 1 assign event for each time the payment task is loaded when a group is not stored in either wp_options or localstorage. There should be as many serve events as you viewed the payment task page.

I've also added ABTest Suite unit tests that will run with `npm run test`

### Changelog

> Add: Add ABTest package.
